spring-integration-security/src/main/java/org/springframework/integration/security/channel/SecurityContextPropagationChannelInterceptor.java

```diff
@@ -0,0 +1,174 @@
+ *
+ * @author Artem Bilan
+ *
+ * @since 4.2
+ */
+public class SecurityContextPropagationChannelInterceptor extends ChannelInterceptorAdapter
+		implements ExecutorChannelInterceptor, BeanFactoryAware {
+
+	public static final String SECURITY_CONTEXT_PROPAGATION_CHANNEL_INTERCEPTOR_BEAN_NAME =
+			"integration" + SecurityContextPropagationChannelInterceptor.class.getName();
+
+	static final ThreadLocal<Deque<SecurityContext>> ORIGINAL_CONTEXT = new ThreadLocal<Deque<SecurityContext>>();
+
+	private final String authenticationHeaderName;
+
+	private volatile MessageBuilderFactory messageBuilderFactory = new DefaultMessageBuilderFactory();
+
+	private volatile boolean messageBuilderFactorySet;
+
+	private BeanFactory beanFactory;
+
+	/**
+	 * Creates a new instance using the header of the name
+	 * {@link SimpMessageHeaderAccessor#USER_HEADER}.
+	 */
+	public SecurityContextPropagationChannelInterceptor() {
+		this(SimpMessageHeaderAccessor.USER_HEADER);
+	}
+
+	/**
+	 * Creates a new instance that uses the specified header to obtain the {@link Authentication}.
+	 * @param authenticationHeaderName the header name to obtain the {@link Authentication}. Cannot be null.
+	 */
+	public SecurityContextPropagationChannelInterceptor(String authenticationHeaderName) {
+		Assert.hasText(authenticationHeaderName, "'authenticationHeaderName' must not be empty.");
+		this.authenticationHeaderName = authenticationHeaderName;
+	}
+
+	@Override
+	public void setBeanFactory(BeanFactory beanFactory) throws BeansException {
+		this.beanFactory = beanFactory;
+	}
+
+	@Override
+	public Message<?> preSend(Message<?> message, MessageChannel channel) {
+		Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
+		if (authentication == null
+				|| message.getHeaders().containsKey(this.authenticationHeaderName)
+				|| isDirectChannel(channel)) {
+			return message;
+		}
+
+		return getMessageBuilderFactory()
+				.fromMessage(message)
+				.setHeader(this.authenticationHeaderName, authentication)
+				.build();
+	}
+
+	private boolean isDirectChannel(MessageChannel channel) {
+		return DirectChannel.class.isAssignableFrom(AopUtils.getTargetClass(channel));
+	}
+
+	@Override
+	public Message<?> beforeHandle(Message<?> message, MessageChannel channel, MessageHandler handler) {
+		setup(message);
+		return message;
+	}
+
+	@Override
+	public Message<?> postReceive(Message<?> message, MessageChannel channel) {
+		if (message != null) {
+			setup(message);
+		}
+		return message;
+	}
+
+	@Override
+	public void afterMessageHandled(Message<?> message, MessageChannel channel, MessageHandler handler, Exception ex) {
+	}
+
+	protected MessageBuilderFactory getMessageBuilderFactory() {
+		if (!this.messageBuilderFactorySet) {
+			if (this.beanFactory != null) {
+				this.messageBuilderFactory = IntegrationUtils.getMessageBuilderFactory(this.beanFactory);
+			}
+			this.messageBuilderFactorySet = true;
+		}
+		return this.messageBuilderFactory;
```

Comment:
?? it will be null if there's no BF...

    if (this.beanFactory != null) {
   			this.messageBuilderFactory =